xdsl/irdl/constraints.py

```diff
@@ -20,38 +21,35 @@ class ConstraintContext:
     Contains the assignment of constraint variables.
     """
 
-    _variables: dict[str, Attribute] = field(default_factory=dict)
+    _variables: ScopedDict[str, Attribute] = field(default_factory=lambda: ScopedDict())
```

Comment:
```suggestion
    _variables: ScopedDict[str, Attribute] = field(default_factory=ScopedDict)
```
Does this work?

Reply:
Unfortunately Pyright isn't happy with this, even though it would work, I have a feeling they have some special handling of the builtin data structures

Reply:
Unfortunate